apps/src/lib/tools/jsdebugger/redux.js

```diff
@@ -0,0 +1,261 @@
+import CommandHistory from './CommandHistory';
+import JSInterpreter from '../../../JSInterpreter';
+import watchedExpressions from '@cdo/apps/redux/watchedExpressions';
+
+const WATCH_TIMER_PERIOD = 250;
+const INITIALIZE = 'jsdebugger/INITIALIZE';
+const ATTACH = 'jsdebugger/ATTACH';
+const DETACH = 'jsdebugger/DETACH';
+const MIRROR_INTERPRETER_STATE = 'jsdebugger/MIRROR_INTERPRETER_STATE';
+const APPEND_LOG = 'jsdebugger/APPEND_LOG';
+const CLEAR_LOG = 'jsdebugger/CLEAR_LOG';
+
+// State model
+const JSDebuggerState = Immutable.Record({
+  jsInterpreter: null,
+  runApp: null,
+  observer: null,
+  watchIntervalId: null,
+  commandHistory: null,
+  logOutput: '',
+  jsInterpreterNextStep: null,
+  jsInterpreterPaused: false,
+});
+
+export function getRoot(state) {
+  return state.jsdebugger;
+}
+
+export function getCommandHistory(state) {
+  return getRoot(state).commandHistory;
+}
+
+export function getJSInterpreter(state) {
+  return getRoot(state).jsInterpreter;
+}
+
+function getRunApp(state) {
+  return getRoot(state).runApp;
+}
+
+export function isPaused(state) {
+  return getRoot(state).jsInterpreterPaused;
+}
+
+function getObserver(state) {
+  return getRoot(state).observer;
+}
+
+function getWatchIntervalId(state) {
+  return getRoot(state).watchIntervalId;
+}
+
+export function isAttached(state) {
+  return !!getJSInterpreter(state);
+}
+
+export function canRunNext(state) {
+  return (
+    isPaused(state) &&
+    getRoot(state).jsInterpreterNextStep === JSInterpreter.StepType.RUN
+  );
+}
+
+export function getLogOutput(state) {
+  return getRoot(state).logOutput;
+}
+
+export const selectors = {
+  getRoot,
+  getCommandHistory,
+  getJSInterpreter,
+  isPaused,
+  isAttached,
+  canRunNext,
+  getLogOutput,
+};
+
+// actions
+export function initialize({runApp}) {
+  return {type: INITIALIZE, runApp};
+}
+
+export function appendLog(output) {
+  return {type: APPEND_LOG, output};
+}
+
+export function clearLog() {
+  return {type: CLEAR_LOG};
+}
+
+function mirrorInterpreterState(jsInterpreter) {
+  return {type: MIRROR_INTERPRETER_STATE, jsInterpreter};
+}
+
+export function attach(jsInterpreter) {
+  return (dispatch, getState) => {
+    const observer = new Observer();
+    observer.observe(
+      jsInterpreter.onNextStepChanged,
+      () => dispatch(mirrorInterpreterState(jsInterpreter))
+    );
+    observer.observe(
+      jsInterpreter.onPause,
+      () => dispatch(togglePause())
+    );
+    observer.observe(
+      jsInterpreter.onExecutionWarning,
+      (output) => dispatch(appendLog(output))
+    );
+
+    const watchIntervalId = setInterval(
+      () => {
+        const jsInterpreter = getJSInterpreter(getState());
+        getState().watchedExpressions.forEach(we => {
+          const currentValue = jsInterpreter.evaluateWatchExpression(we.get('expression'));
+          dispatch(updateWatchExpressions(we.get('expression'), currentValue));
+        });
+      },
+      WATCH_TIMER_PERIOD
+    );
```

Comment:
I would like to revisit this `watchIntervalId` stuff. Seems like it should just belong in the `watchedExpressions` code? I think the only reason why it is in `JsDebuggerUI` is because that's where there was a reference to the `JSInterpreter` instance which is needed to evaluate the watch expression.